SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc

```diff
@@ -119,39 +123,53 @@ void HGCalSiNoiseMap::setGeometry(const CaloSubdetectorGeometry *hgcGeom, GainRa
 }
 
 //
-const HGCalSiNoiseMap::SiCellOpCharacteristicsCore HGCalSiNoiseMap::getSiCellOpCharacteristicsCore(
-    const HGCSiliconDetId &cellId, GainRange_t gain, int aimMIPtoADC) {
+template <typename T>
+const typename HGCalSiNoiseMap<T>::SiCellOpCharacteristicsCore HGCalSiNoiseMap<T>::getSiCellOpCharacteristicsCore(
+    const T &cellId, GainRange_t gain, int aimMIPtoADC) {
   //re-compute
   if (!activateCachedOp_)
     return getSiCellOpCharacteristics(cellId, gain, aimMIPtoADC).core;
 
-  //re-use from cache
-  HGCSiliconDetId posCellId(cellId.subdet(),
-                            1,
-                            cellId.type(),
-                            cellId.layer(),
-                            cellId.waferU(),
-                            cellId.waferV(),
-                            cellId.cellU(),
-                            cellId.cellV());
-  uint32_t key(posCellId.rawId());
+  uint32_t key;
+
+  if (cellId.det() == DetId::Forward && cellId.subdetId() == ForwardSubdetector::HFNose) {
```

Comment:
I'd remove the `if`, implement the default for HGCAL and specialize the one for the `HFNose`.

Reply:
Ciao @mariadalfonso 
I really meant to code a `template specialization` using the specific type you need, i.e., `HFNoseDetId`. In this case, no `if` is required anywhere in the code, since 2 difference functions, with different signatures, will be generated, each containing the correct code to handle the specific type.
I assume the code calling this method will already know the correct `type`.

Reply:
@rovere 
The thing is that the constructor of the HFNoseDetId and HGCSiliconDetId are different, so cannot construct a simple object of T , that will get to use on the fly. So I keep the if because of this.

Reply:
Ciao @mariadalfonso I'm not sure I follow your reasoning. Those pieces of code, that depend on the specific types, will end up in different methods: you can use the most appropriate one, depending on the type.
The whole point of templates is, also, to avoid at run-time to ask every time `who-are-you` if you know already the type.

Reply:
seems that you prefer duplicate entire methods, rather than putting the if condition. Sure can be done.

Reply:
uhm, the duplication cannot done that easily
getSiCellOpCharacteristicsCore get called in the HGCDigitizerBase.cc where one "know the type" and inside the HGCalSiNoiseMap::setGeometry where one "doesn't know the type".

the best would be to do this key in a more general way.

Reply:
I think you know the type also in that case, no? See [here](https://github.com/cms-sw/cmssw/pull/31981/files#diff-4297dad63e271c9b557a45d4fa858ef38e0e63bb43611c49f816be2540138a21R95-R96): you call `setGeometry` on two different objects, each of which should know the correct type.

The problem is that the way in which everything is handled is fragile, to say the least.

What is this [none-sense](https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc#L827-L832)? 
I guess that the whole trick is handled there, where you get correct geometry and the correct list of valid detid, for each case.

Reply:
@rovere actually the key is the raw_id, there is no point to build again the object
so now there are no cloned methods nor IF and might be faster